datamol/graph.py

```diff
@@ -264,35 +267,71 @@ def reorder_mol_from_template(
             logger.warning("No match was found")
         return None
 
-    # In case we want to allow ambiguous match of hydrogens
-    if (len(matches) > 1) and allow_ambiguous_hs_only and (not allow_ambiguous_match):
-        first_keys = list(matches[0].keys())
-        all_hs_mismatch = True
-        for this_match in matches:
-            this_keys = list(this_match.keys())
-            keys_mismatch = [
-                ii for ii in range(len(this_keys)) if (first_keys[ii] != this_keys[ii])
+    if len(matches) > 1:
+        # In case we want to allow ambiguous match of hydrogens
+        if allow_ambiguous_match == "hs-only":
+            first_keys = list(matches[0].keys())
+            all_hs_mismatch = True
+            for this_match in matches:
+                this_keys = list(this_match.keys())
+                keys_mismatch = [
+                    ii for ii in range(len(this_keys)) if (first_keys[ii] != this_keys[ii])
+                ]
+                atoms_mismatch = [mol.GetAtomWithIdx(key).GetAtomicNum() for key in keys_mismatch]
+                all_hs = all([atom == 1 for atom in atoms_mismatch])
+                if not all_hs:
+                    all_hs_mismatch = False
+                    break
+            if all_hs_mismatch:
+                matches = matches[0:1]
+            else:
+                if verbose:
+                    logger.warning(
+                        f"{len(matches)} matches were found, ordering is ambiguous, even when ignoring hydrogens"
+                    )
+                return None
+
+        # Compute the number of atoms and bonds mismatch, and select the one with the least mismatch
+        if (allow_ambiguous_match in ["best", "best-first"]) and not (
+            enforce_atomic_num and enforce_bond_type
+        ):
+            num_mismatches = []
+            for this_match in matches:
+                num_atoms_mismatch = 0
+
+                # Get the number of atomic mismatch
+                for key, val in this_match.items():
+                    atom1 = mol.GetAtomWithIdx(val).GetAtomicNum()
+                    atom2 = mol_template.GetAtomWithIdx(key).GetAtomicNum()
+                    num_atoms_mismatch += atom1 != atom2
+
+                # Get the number of bond mismatch
+                num_bonds_type_mismatch, num_bonds_stereo_mismatch = 0, 0
+                for bond1 in mol_template.GetBonds():
```

Comment:
Would need to dig further, but I think you shouldn't be iterating over all bonds. Instead you should focus on the bonds where atoms in the match participates in both molecules

Reply:
Reason is `this_match[xxx]` can be undefined at line 312.  But maybe you have covered that case and I missed it.

Reply:
What do you mean by `you should focus on the bonds where atoms in the match participates`? What does participate mean?

Reply:
Discard this, since you don't allow partial match at all (you either have a bijection or None).